Failed t/local/10_rand.t

Hi!

```
t/local/10_rand.t ...................... 1/52 
#   Failed test 'RAND_file_name returns defined value'
#   at t/local/10_rand.t line 18.
#          got: undef
#     expected: anything else
Use of uninitialized value $file_name in concatenation (.) or string at t/local/10_rand.t line 19.
# Looks like you failed 1 test of 52.
t/local/10_rand.t ...................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/52 subtests 

```
full log [here](http://pastebin.calculate-linux.ru/ru/show/128183)

---

[@chrisnovakovic edit to add the full log from the link above, in case it goes missing:]

```
t/handle/local/05_use.t ................ ok   
t/local/01_pod.t ....................... skipped: Test::Pod 1.00 required for testing POD
t/local/02_pod_coverage.t .............. skipped: these tests are for only for release candidate testing. Enable with RELEASE_TESTING=1
t/local/03_use.t ....................... 1/1 # 
# Testing Net::SSLeay 1.88
# 
# Perl information:
#   Version:         '5.026001'
#   Executable path: '/home/guest/perl5/perlbrew/perls/perl-5.26.1/bin/perl'
# 
# libssl information:
#   SSLEAY_VERSION:      'OpenSSL 1.1.1d  10 Sep 2019'
#   SSLEAY_CFLAGS:       'compiler: x86_64-pc-linux-gnu-gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -march=native -O2 -pipe -fno-strict-aliasing -Wa,--noexecstack -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DZLIB -DNDEBUG  -DOPENSSL_NO_BUF_FREELISTS'
#   SSLEAY_BUILT_ON:     'built on: Thu Oct 17 13:30:14 2019 UTC'
#   SSLEAY_PLATFORM:     'platform: linux-x86_64'
#   SSLEAY_DIR:          'OPENSSLDIR: "/etc/ssl"'
#   OPENSSL_ENGINES_DIR: 'ENGINESDIR: "/usr/lib64/engines-1.1"'
t/local/03_use.t ....................... ok   
t/local/04_basic.t ..................... ok     
t/local/05_passwd_cb.t ................. ok     
t/local/06_tcpecho.t ................... ok   
t/local/07_sslecho.t ................... ok       
t/local/08_pipe.t ...................... ok     
t/local/09_ctx_new.t ................... ok     
t/local/10_rand.t ...................... 1/52 
#   Failed test 'RAND_file_name returns defined value'
#   at t/local/10_rand.t line 18.
#          got: undef
#     expected: anything else
Use of uninitialized value $file_name in concatenation (.) or string at t/local/10_rand.t line 19.
# Looks like you failed 1 test of 52.
t/local/10_rand.t ...................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/52 subtests 
t/local/11_read.t ...................... ok     
t/local/15_bio.t ....................... ok   
t/local/20_autoload.t .................. ok       
t/local/21_constants.t ................. ok       
t/local/30_error.t ..................... ok     
t/local/31_rsa_generate_key.t .......... ok     
t/local/32_x509_get_cert_info.t ........ ok         
t/local/33_x509_create_cert.t .......... ok       
t/local/34_x509_crl.t .................. ok     
t/local/35_ephemeral.t ................. skipped: libressl and OpenSSL 1.1 removed support for ephemeral/temporary RSA private keys
t/local/36_verify.t .................... ok       
t/local/37_asn1_time.t ................. ok     
t/local/38_priv-key.t .................. ok     
t/local/39_pkcs12.t .................... ok     
t/local/40_npn_support.t ............... ok   
t/local/41_alpn_support.t .............. ok   
t/local/42_info_callback.t ............. ok   
t/local/43_misc_functions.t ............ ok     
t/local/44_sess.t ...................... ok     
t/local/45_exporter.t .................. ok     
t/local/50_digest.t .................... ok       
t/local/61_threads-cb-crash.t .......... skipped: your perl is not compiled with ithreads
t/local/62_threads-ctx_new-deadlock.t .. skipped: your perl is not compiled with ithreads
t/local/63_ec_key_generate_key.t ....... ok   
t/local/64_ticket_sharing.t ............ ok     
t/local/65_security_level.t ............ ok     
t/local/65_ticket_sharing_2.t .......... ok   
t/local/66_curves.t .................... ok   
t/local/kwalitee.t ..................... skipped: these tests are for only for release candidate testing. Enable with RELEASE_TESTING=1

Test Summary Report
-------------------
t/local/10_rand.t                    (Wstat: 256 Tests: 52 Failed: 1)
  Failed test:  3
  Non-zero exit status: 1
Files=40, Tests=3039,  6 wallclock secs ( 0.40 usr  0.05 sys +  4.44 cusr  0.31 csys =  5.20 CPU)
Result: FAIL
Failed 1/40 test programs. 1/3039 subtests failed.
make: *** [Makefile:1081: test_dynamic] Ошибка 255
FAIL
```

--- a/Changes
+++ b/Changes
@@ -22,11 +22,16 @@ Revision history for Perl extension Net::SSLeay.
 	- Expose SSL_CTX_set_keylog_callback and
 	  SSL_CTX_get_keylog_callback available with OpenSSL 1.1.1pre1
 	  and later.
-	- Enhance 10_rand.t tests so that they don't depend on the
-	  runtime environment variables, such as HOME and
-	  RANDFILE. Address differences between OpenSSL versions. Note
-	  in SSLeay.pod that RAND_file_name() can return undef with
-	  LibreSSL and recent OpenSSL versions. Fixes GH-152.
+	- Enhance 10_rand.t RAND_file_name tests: tests are no longer
+	  affected by the runtime environment variables, HOME and
+	  RANDFILE. These variables are insted controlled by the tests
+	  with local %ENV. Problems related to RAND_file_name were
+	  discussed in Github issues GH-152, and there might still be
+	  cases when, for example, setuid is used because of OpenSSL's
+	  use of glibc secure_getenv() and related functions. Address
+	  RAND_file_name differences between OpenSSL versions. Note in
+	  SSLeay.pod that RAND_file_name() can return undef with
+	  LibreSSL and recent OpenSSL versions.
 
 1.91_01 2021-10-24
 	- Correct X509_STORE_CTX_init() return value to integer. Previous

--- a/Changes
+++ b/Changes
@@ -22,6 +22,11 @@ Revision history for Perl extension Net::SSLeay.
 	- Expose SSL_CTX_set_keylog_callback and
 	  SSL_CTX_get_keylog_callback available with OpenSSL 1.1.1pre1
 	  and later.
+	- Enhance 10_rand.t tests so that they don't depend on the
+	  runtime environment variables, such as HOME and
+	  RANDFILE. Address differences between OpenSSL versions. Note
+	  in SSLeay.pod that RAND_file_name() can return undef with
+	  LibreSSL and recent OpenSSL versions. Fixes GH-152.
 
 1.91_01 2021-10-24
 	- Correct X509_STORE_CTX_init() return value to integer. Previous

--- a/Changes
+++ b/Changes
@@ -26,7 +26,7 @@ Revision history for Perl extension Net::SSLeay.
 	  affected by the runtime environment variables, HOME and
 	  RANDFILE. These variables are insted controlled by the tests
 	  with local %ENV. Problems related to RAND_file_name were
-	  discussed in Github issues GH-152, and there might still be
+	  discussed in Github issue GH-152, and there might still be
 	  cases when, for example, setuid is used because of OpenSSL's
 	  use of glibc secure_getenv() and related functions. Address
 	  RAND_file_name differences between OpenSSL versions. Note in

--- a/SSLeay.xs
+++ b/SSLeay.xs
@@ -3258,7 +3258,7 @@ RAND_file_name(num)
     PREINIT:
         char *buf;
     CODE:
-        New(0, buf, num, char);
+        Newxz(buf, num, char);
         if (!RAND_file_name(buf, num)) {
             Safefree(buf);
             XSRETURN_UNDEF;

--- a/lib/Net/SSLeay.pod
+++ b/lib/Net/SSLeay.pod
@@ -5423,7 +5423,9 @@ Generates a default path for the random seed file.
  my $file = Net::SSLeay::RAND_file_name($num);
  # $num - maximum size of returned file name
  #
- # returns: string with file name on success, '' (empty string) on failure
+ # returns: string with file name on success, '' (empty string) or undef on failure
+
+LibreSSL and OpenSSL 1.1.0a and later return undef when, for example, $num is not large enough to hold the filename.
 
 Check openssl doc L<http://www.openssl.org/docs/crypto/RAND_load_file.html|http://www.openssl.org/docs/crypto/RAND_load_file.html>
 

--- a/t/local/10_rand.t
+++ b/t/local/10_rand.t
@@ -61,6 +61,10 @@ sub test_rand_file_name_openssl
     local %ENV = %ENV;
     delete $ENV{RANDFILE};
 
+    # NOTE: If there are tests failures, are you using some type of
+    # setuid environment? If so, this may affect usability of
+    # environment variables.
+
     $ENV{HOME} = '/nosuchdir-1/home';
     $file_name = Net::SSLeay::RAND_file_name(300);
     if (Net::SSLeay::SSLeay() >= 0x10100006 && Net::SSLeay::SSLeay() <= 0x1010000f)
@@ -74,7 +78,14 @@ sub test_rand_file_name_openssl
     my $randfile = '/nosuchdir-2/randfile';
     $ENV{RANDFILE} = $randfile;
     $file_name = Net::SSLeay::RAND_file_name(300);
-    is($file_name, $randfile, "RAND_file_name return value '$file_name' is RANDFILE environment value");
+    if (Net::SSLeay::SSLeay() < 0x1010001f) {
+	# On Windows, and possibly other non-Unix systems, 1.0.2
+	# series and earlier did not honour RANDFILE. 1.1.0a is an
+	# educated guess when it starts working with all platforms.
+	isnt($file_name, q{}, "RAND_file_name returns non-empty string when RANDFILE is set: $file_name");
+    } else {
+	is($file_name, $randfile, "RAND_file_name return value '$file_name' is RANDFILE environment value");
+    }
 
     # RANDFILE is longer than 2 octets. OpenSSL 1.1.0a and later
     # return undef with short buffer

--- a/t/local/10_rand.t
+++ b/t/local/10_rand.t
@@ -54,14 +54,15 @@ sub test_rand_file_name_libressl
 # RAND_file_name return value. Note: we override environment variables
 # temporarily because some environments do not have HOME set or may
 # already have RANDFILE set. We do not try to trigger a failure which
-# happens if there's no HOME nor RANDFILE. This is because
+# happens if there's no HOME nor RANDFILE in order to keep the test
+# from becoming overly complicated.
 sub test_rand_file_name_openssl
 {
     my $file_name;
     local %ENV = %ENV;
     delete $ENV{RANDFILE};
 
-    # NOTE: If there are tests failures, are you using some type of
+    # NOTE: If there are test failures, are you using some type of
     # setuid environment? If so, this may affect usability of
     # environment variables.
 

--- a/t/local/10_rand.t
+++ b/t/local/10_rand.t
@@ -3,19 +3,19 @@
 use lib 'inc';
 
 use Net::SSLeay;
-use Test::Net::SSLeay qw( data_file_path initialise_libssl );
+use Test::Net::SSLeay qw( data_file_path initialise_libssl is_libressl );
 
-plan tests => 52;
+plan tests => 53;
 
 initialise_libssl();
 
 is(Net::SSLeay::RAND_status(), 1, 'RAND_status');
 is(Net::SSLeay::RAND_poll(), 1, 'RAND_poll');
 
-# RAND_file_name
-my $file_name = Net::SSLeay::RAND_file_name(300);
-isnt($file_name, undef, 'RAND_file_name returns defined value');
-isnt($file_name, "", "RAND_file_name returns non-empty string: $file_name");
+# RAND_file_name has significant differences between the two libraries
+is_libressl() ?
+    test_rand_file_name_libressl() :
+    test_rand_file_name_openssl();
 
 # RAND_load_file
 my $binary_file      = data_file_path('binary-test.file');
@@ -34,6 +34,59 @@ if (Net::SSLeay::constant("LIBRESSL_VERSION_NUMBER"))
 test_rand_bytes();
 
 exit(0);
+
+# With LibreSSL RAND_file_name is expected to always succeed as long
+# as the buffer size is large enough. Their manual states that it's
+# implemented for API compatibility only and its use is discouraged.
+sub test_rand_file_name_libressl
+{
+    my $file_name = Net::SSLeay::RAND_file_name(300);
+    isnt($file_name, undef, 'RAND_file_name returns defined value');
+    isnt($file_name, q{}, "RAND_file_name returns non-empty string: $file_name");
+
+    $file_name = Net::SSLeay::RAND_file_name(2);
+    is($file_name, undef, "RAND_file_name return value is undef with too short buffer");
+
+    return;
+}
+
+# With OpenSSL there are a number of options that affect
+# RAND_file_name return value. Note: we override environment variables
+# temporarily because some environments do not have HOME set or may
+# already have RANDFILE set. We do not try to trigger a failure which
+# happens if there's no HOME nor RANDFILE. This is because
+sub test_rand_file_name_openssl
+{
+    my $file_name;
+    local %ENV = %ENV;
+    delete $ENV{RANDFILE};
+
+    $ENV{HOME} = '/nosuchdir-1/home';
+    $file_name = Net::SSLeay::RAND_file_name(300);
+    if (Net::SSLeay::SSLeay() >= 0x10100006 && Net::SSLeay::SSLeay() <= 0x1010000f)
+    {
+	# This was broken starting with 1.0.0-pre6 and fixed after 1.0.0
+	is($file_name, q{}, "RAND_file_name return value is empty and doesn't include '.rnd'");
+    } else {
+	like($file_name, qr/\.rnd/s, "RAND_file_name return value '$file_name' includes '.rnd'");
+    }
+
+    my $randfile = '/nosuchdir-2/randfile';
+    $ENV{RANDFILE} = $randfile;
+    $file_name = Net::SSLeay::RAND_file_name(300);
+    is($file_name, $randfile, "RAND_file_name return value '$file_name' is RANDFILE environment value");
+
+    # RANDFILE is longer than 2 octets. OpenSSL 1.1.0a and later
+    # return undef with short buffer
+    $file_name = Net::SSLeay::RAND_file_name(2);
+    if (Net::SSLeay::SSLeay() < 0x1010001f) {
+	is($file_name, q{}, "RAND_file_name return value is empty string with too short buffer");
+    } else {
+	is($file_name, undef, "RAND_file_name return value is undef with too short buffer");
+    }
+
+    return;
+}
 
 sub test_rand_bytes
 {

